File: grpc/hello-grpc-python/client/protoClient.py
# encoding: utf-8
import logging
import os
import random
import time

from conn import connection
from landing_pb2 import landing_pb2
from landing_pb2 import landing_pb2_grpc

logger = logging.getLogger('grpc-client')
logger.setLevel(logging.INFO)
console = logging.StreamHandler()
console.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s [%(levelname)s] - %(message)s')
console.setFormatter(formatter)
logger.addHandler(console)


def talk(stub):
    request = landing_pb2.TalkRequest(data="0", meta="PYTHON")
    logger.info("Talk data:%s,meta:%s", request.data, request.meta)
    metadata = (("k1", "v1"),
                ("k2", "v2"),)
    response = stub.Talk(request=request, metadata=metadata)
    print_response("Talk", response)


def talk_one_answer_more(stub):
    request = landing_pb2.TalkRequest(data="0,1,2", meta="PYTHON")
    logger.info("TalkOneAnswerMore data:%s,meta:%s", request.data, request.meta)
    metadata = (("k1", "v1"),
                ("k2", "v2"),)
    responses = stub.TalkOneAnswerMore(request=request, metadata=metadata)
    for response in responses:
        print_response("TalkOneAnswerMore", response)


def talk_more_answer_one(stub):
    request_iterator = generate_request("TalkMoreAnswerOne")
    metadata = (("k1", "v1"),
                ("k2", "v2"),)
    response_summary = stub.TalkMoreAnswerOne(request_iterator=request_iterator, metadata=metadata)
    print_response("TalkMoreAnswerOne", response_summary)


def talk_bidirectional(stub):
    request_iterator = generate_request("TalkBidirectional")
    metadata = (("k1", "v1"),
                ("k2", "v2"),)
    responses = stub.TalkBidirectional(request_iterator=request_iterator, metadata=metadata)
    for response in responses:
        print_response("TalkBidirectional", response)


def grpc_server():
    server = os.getenv("GRPC_SERVER")
    if server:
        return server
    else:
        return "localhost"


def random_id(end):
    return str(random.randint(0, end))


def generate_request(method_name):
    for _ in range(0, 3):
        request = landing_pb2.TalkRequest(data=random_id(5), meta="PYTHON")
        logger.info("%s data:%s,meta:%s", method_name, request.data, request.meta)
        yield request
        time.sleep(random.uniform(0.5, 1.5))


def print_response(method_name, response):
    for result in response.results:
        kv = result.kv
        logger.info("%s [%d] %d [%s %s %s,%s:%s]", method_name,
                    response.status, result.id, kv["meta"], result.type, kv["id"], kv["idx"], kv["data"])


def run():
    address = grpc_server() + ":9996"
    channel = connection.build_channel(address)
    stub = landing_pb2_grpc.LandingServiceStub(channel)
    logger.info("Unary RPC")
    talk(stub)
    logger.info("Server streaming RPC")
    talk_one_answer_more(stub)
    logger.info("Client streaming RPC")
    talk_more_answer_one(stub)
    logger.info("Bidirectional streaming RPC")
    talk_bidirectional(stub)


if __name__ == '__main__':
    run()

File: grpc/hello-grpc-kotlin/client/src/main/kotlin/org/feuyeux/grpc/ProtoClient.kt
package org.feuyeux.grpc

import io.grpc.ManagedChannel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow
import org.apache.logging.log4j.kotlin.logger
import org.feuyeux.grpc.conn.Connection
import org.feuyeux.grpc.proto.LandingServiceGrpcKt.LandingServiceCoroutineStub
import org.feuyeux.grpc.proto.TalkRequest
import org.feuyeux.grpc.proto.TalkResponse
import java.io.Closeable
import java.util.concurrent.TimeUnit

private fun getGrcServer(): String {
    return System.getenv("GRPC_SERVER") ?: return "localhost"
}

suspend fun main() {
    val channel = Connection.getChannel(HeaderClientInterceptor())
    ProtoClient(channel).use {
        var talkRequest: TalkRequest = TalkRequest.newBuilder()
            .setMeta("KOTLIN")
            .setData("0")
            .build()
        val response1: TalkResponse = it.talk(talkRequest)
        it.printResponse(response1)

        talkRequest = TalkRequest.newBuilder()
            .setMeta("KOTLIN")
            .setData("0,1,2")
            .build()
        val responseFlow1 = it.talkOneAnswerMore(talkRequest)
        responseFlow1.collect { response2 ->
            it.printResponse(response2)
        }

        val requests = listOf(
            TalkRequest.newBuilder()
                .setMeta("KOTLIN")
                .setData((0..5).random().toString())
                .build(),
            TalkRequest.newBuilder()
                .setMeta("KOTLIN")
                .setData((0..5).random().toString())
                .build(),
            TalkRequest.newBuilder()
                .setMeta("KOTLIN")
                .setData((0..5).random().toString())
                .build()
        )
        val response3: TalkResponse = it.talkMoreAnswerOne(requests)
        it.printResponse(response3)

        val responseFlow2 = it.talkBidirectional(requests)
        responseFlow2.collect { response4 ->
            it.printResponse(response4)
        }
        // TimeUnit.SECONDS.sleep(5)
    }
}

class ProtoClient(private val channel: ManagedChannel) : Closeable {
    private val logger = logger()
    private val stub = LandingServiceCoroutineStub(channel)

    override fun close() {
        channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
        logger.info("Done")
    }

    suspend fun talk(talkRequest: TalkRequest): TalkResponse {
        logger.info("Unary RPC")
        return stub.talk(talkRequest)
    }

    fun talkOneAnswerMore(talkRequest: TalkRequest): Flow<TalkResponse> {
        logger.info("Server streaming RPC")
        return stub.talkOneAnswerMore(talkRequest)
    }

    suspend fun talkMoreAnswerOne(talkRequests: List<TalkRequest>): TalkResponse {
        logger.info("Client streaming RPC")
        fun listToFlow(rs: List<TalkRequest>): Flow<TalkRequest> = flow {
            for (r in rs) {
                emit(r)
                delay(timeMillis = (5L..10L).random())
            }
        }
        return stub.talkMoreAnswerOne(listToFlow(talkRequests))
    }

    suspend fun talkBidirectional(talkRequests: List<TalkRequest>): Flow<TalkResponse> {
        logger.info("Bidirectional streaming RPC")
        fun listToFlow(rs: List<TalkRequest>): Flow<TalkRequest> = flow {
            for (r in rs) {
                emit(r)
                delay(timeMillis = 5L)
            }
        }
        return stub.talkBidirectional(listToFlow(talkRequests))
    }

    internal fun printResponse(response: TalkResponse) {
        response.resultsList.forEach { result ->
            val kv = result.kvMap
            logger.info("${response.status} ${result.id} [${kv["meta"]} ${result.type} ${kv["id"]}, ${kv["idx"]}:${kv["data"]}]")
        }
    }
}

